design/20230726-cel-policy.md

```diff
@@ -0,0 +1,369 @@
+
+<!--
+For example, will cert-manager `CustomResourceDefinition`s increase in size, will there be more `Secret`s or `CertificateRequest`s created?
+-->
+
+### Will enabling / using this feature result in significant increase of resource usage? (CPU, RAM...)
+
+<!--
+For example, will implementing this feature result in more objects being cache thus increasing memory consumption?
+-->
+
+## Drawbacks
+
+<!--
+Why should this proposal _not_ be implemented?
+-->
+
+## Alternatives
+
+<!--
+What other approaches did you consider, and why did you rule them out? These do
+not need to be as detailed as the proposal, but should include enough
+information to express the idea and why it was not acceptable.
+-->
+
+### CEL expressions with full decoded CSR variable
+
+This alternative was suggested by @inteon on Slack.
+
+The idea here is to provide the policy author with a CEL context variable containing the decoded CSR replacing the
+`self` variable in the proposed design. Since the expressions now get access to the full CSR, the expressions will have
+to be "promoted" in the CertificateRequestPolicy API. The expression will have to return a bool indicating approved or
+denied. An example policy with this alternative API might look like (example);
+
+```yaml
+spec:
+  expressions:
+    - >-
+      has(csr.dnsNames) &&
+      csr.dnsNames.all(dnsName,
+      dnsName.endsWith(cr.namespace + '.svc') ||
+      dnsName.endsWith(cr.namespace + '.svc.cluster.local'))
+```
+
+Pros:
+
+- Extremely flexible. I.e. allows for CSR cross-attribute validation.
+- Some would find this API more expressive and better describe the intentions.
+
+Cons:
+
+- More complex expressions that will require more in-depth knowledge of CEL.
+- Will create a "split" API, since you would probably use either `expressions` or `allowed`/`constraints`.
+  This could increase the maintenance burden and source to subtle bugs. Breaking the API could be an option to mend this.
+- Will probably require custom CEL functions to fulfill the requirements in approver-policy. This is the main reason for
+  recommending this design, and will be elaborated in the following.
+
+Let me start with a quote from the
+[approver-policy documentation](https://cert-manager.io/docs/projects/approver-policy/#allowed):
+
+> Allowed is the block that defines attributes that match against the corresponding attribute in the request.
+> A request is permitted by the policy if the request omits an allowed attribute, but will deny the request if
+> it contains an attribute which is not present in the allowed block.
+
+The last sentence describes the requirement that could need custom CEL function(s) to allow the user to specify allowed
+CSR attributes. We cannot expect the user to include validations for **all** CSR attributes in the expression(s). So I
+think we at least need a function to list allowed CSR attribute usage.
```

Comment:
I have now looked at the source code, and approver-policy only evaluates a policy against the CSR attributes that approver-policy supports. So you are correct that it might be considered too permissive in this sense. I suggest creating an issue for this if it is something that we should attempt to fix. Changing it would be breaking tho. 😉